### PR TITLE
[Plugins] skeleton for events correlation engine plugin

### PR DESCRIPTION
**Pull request by sbcd90**
### Description
Initial pr for defining skeleton of events correlation engine plugin. RFC is [here](https://github.com/opensearch-project/OpenSearch/issues/6779)

### Issues Resolved
The 1st issue in the [META issue](https://github.com/opensearch-project/OpenSearch/issues/6854) will be resolved.

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sbcd90** [comment]
the `Gradle check` failed due to `test failures`.

```
Tests with failures:
 - org.opensearch.snapshots.SearchableSnapshotIT.testCacheFilesAreClosedAfterUse
```

@nknize @getsaurabh02 kindly have a look.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestSearchCorrelationAction.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.opensearch.plugin.correlation.events.resthandler;
+
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.RestRequest;
+
+import java.io.IOException;
+
+public class RestSearchCorrelationAction extends BaseRestHandler {
```
Comment (getsaurabh02) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestSearchCorrelationAction.java:17]: nit : Should it be ``RestSearchCorrelatedEventsAction`` to keep it consistent with the related Actions? I am open to both

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/rules/action/IndexCorrelationRuleAction.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.correlation.rules.action;
+
+import org.opensearch.action.ActionType;
+
+public class IndexCorrelationRuleAction extends ActionType<IndexCorrelationRuleResponse> {
```
Comment (getsaurabh02) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/rules/action/IndexCorrelationRuleAction.java:13]: We will also need to corresponding ``GetCorrelationRuleAction`` along with its REST and Transport implementations?

**Review by getsaurabh02 — COMMENTED** [review]
Thanks @sbcd90 for putting up this frame. Looks a good start to me and added few minor comments. 
I see we will have other basic Actions components coming in soon such as to handle  ``Correlation Joins`` or ``Correlation Rule Parsing``, which we can add in subsequent PRs too.

**Comment by getsaurabh02** [comment]
Can we also add some java docs to new Actions especially those bring added at the Transport and Rest layers for developer productivity.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestSearchCorrelationAction.java (hunk 1 above)
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestSearchCorrelationAction.java]: renamed it.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/CorrelationConstants.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.correlation.core;
+
+public class CorrelationConstants {
+    public static final String METHOD_PARAMETER_M = "m";
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/CorrelationConstants.java]: removed the variable as this is just the skeleton plugin.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/rules/action/IndexCorrelationRuleAction.java (hunk 2 above)
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/rules/action/IndexCorrelationRuleAction.java:13]: will add more actions with individual prs.

**Comment by sbcd90** [comment]
> Can we also add some java docs to new Actions especially those bring added at the Transport and Rest layers for developer productivity.

As this is just the skeleton, i haven't added any Javadoc as there is mostly no logic. will add code & javadoc for it in subsequent prs.

**Comment by getsaurabh02** [comment]
> > Can we also add some java docs to new Actions especially those bring added at the Transport and Rest layers for developer productivity.
> 
> @getsaurabh02 As this is just the skeleton, i haven't added any Javadoc as there is mostly no logic. will add code & javadoc for it in subsequent prs.

The skeleton itself can have the java doc stating the intent behind. I am fine if we are planning to add them with the code.

**Review by andrross — COMMENTED** [review]
What do you think about moving this to the [sandbox](https://github.com/opensearch-project/OpenSearch/blob/4511354122f8b423bbb78d95b7ed6d699450a9b1/DEVELOPER_GUIDE.md#sandbox) while it is under development?

I definitely don't want to stand in the way of incremental development, and I certainly don't expect this whole feature to shipped in a single PR, but I can't really provide a meaningful review of empty classes. Is there no way to structure this such that you're delivering sections of functional and unit tested code in each PR?

**Pull request closed without merging**